src/utils/cuda/cuda_utils.cpp

```diff
@@ -0,0 +1,363 @@
+
+    nixl_status_t
+    memCtxImpl::queryVmm(const void *address, memory_t &type, int &id)
+    {
+        nixl_status_t ret = NIXL_ERR_NOT_FOUND;
+
+#if HAVE_CUMEMRETAINALLOCATIONHANDLE
+        CUmemAllocationProp prop = {};
+        CUmemGenericAllocationHandle alloc_handle;
+
+        /* Check if memory is allocated using VMM API and see if host memory needs
+        * to be treated as pinned device memory */
+        auto result = cuMemRetainAllocationHandle(&alloc_handle, (void*)address);
+        if (result != CUDA_SUCCESS) {
+            NIXL_TRACE << "cuMemRetainAllocationHandle() failed. result = "
+                    << result;
+            return NIXL_ERR_NOT_FOUND;
+        }
+        // TODO: set the call to cuMemRelease when leaving the scope to avoid GOTO
+
+        result = cuMemGetAllocationPropertiesFromHandle(&prop, alloc_handle);
+        if (result != CUDA_SUCCESS) {
+            NIXL_DEBUG << "cuMemGetAllocationPropertiesFromHandle() failed. result = "
+                    << result;
+            ret = NIXL_ERR_UNKNOWN;
+            goto err;
+        }
+
+        id = prop.location.id;
+        switch (prop.location.type) {
+        case CU_MEM_LOCATION_TYPE_DEVICE:
+            type = MEM_VMM_DEV;
+            ret = NIXL_SUCCESS;
+            break;
+        default:
+            NIXL_DEBUG << "Unsupported VMM memory type: " << prop.location.type;
+            ret = NIXL_ERR_INVALID_PARAM;
+            goto err;
+        }
+
+    err:
+        result = cuMemRelease(alloc_handle);
+        if (CUDA_SUCCESS != result) {
+            NIXL_DEBUG << "cuMemRelease() failed. result = "
+                    << result;
+            if (NIXL_SUCCESS == ret) {
+                ret = NIXL_ERR_UNKNOWN;
+            }
+        }
+
+#endif
+        return ret;
+    }
+
+    nixl_status_t
+    memCtxImpl::retainVmmCudaCtx(int id, CUcontext &newCtx) const
+    {
+        CUdevice device;
+
+        auto result = cuDeviceGet(&device, id);
+        if (result != CUDA_SUCCESS) {
+            NIXL_ERROR << "cuDeviceGet() failed. result = " << result;
+            return NIXL_ERR_UNKNOWN;
+        }
+
+        unsigned int flags;
+        int active;
+        result = cuDevicePrimaryCtxGetState(device, &flags, &active);
+        if (result != CUDA_SUCCESS) {
+            NIXL_ERROR << "cuDevicePrimaryCtxGetState() failed. result = "
+                    << result;
+            return NIXL_ERR_UNKNOWN;
+        }
+
+        if (!active) {
+            NIXL_ERROR << "No active context found for CUDA device " << id;
+            return NIXL_ERR_INVALID_PARAM;
+        }
+
+        result = cuDevicePrimaryCtxRetain(&newCtx, device);
+        if (result != CUDA_SUCCESS) {
+            NIXL_ERROR << "cuDevicePrimaryCtxRetain() failed. result = "
+                    << result;
+            return NIXL_ERR_UNKNOWN;
+        }
+
+        return NIXL_SUCCESS;
+    }
+
+    nixl_status_t
+    memCtxImpl::releaseVmmCudaCtx(int id) const
```

Comment:
This method should be void, and print warnings instead of errors (NIXL_WARN). Because the application can continue running with unreleased primary context. And the error from dtor can be removed since it just duplicates the message already printed by the function.

Reply:
should be fixed in  ba3eb63dc41985235d7f3d51ef15239ab4c9b21e